core/src/main/java/org/apache/brooklyn/core/typereg/RegisteredTypes.java

```diff
@@ -66,180 +94,131 @@ public static RegisteredType of(CatalogItem<?, ?> item) {
         type.deprecated = item.isDeprecated();
 
         // TODO
-        // javaType, specType, registeredTypeName ...
-        // tags ?
+        // probably not: javaType, specType, registeredTypeName ...
+        // maybe: tags ?
         return type;
     }
 
-    /** Visitor adapter which can be used to ensure all kinds are supported */
-    public static abstract class RegisteredTypeKindVisitor<T> {
-        public T visit(RegisteredType type) {
-            if (type==null) throw new NullPointerException("Registered type must not be null");
-            if (type instanceof RegisteredSpecType) {
-                return visitSpec((RegisteredSpecType)type);
-            }
-            // others go here
-            throw new IllegalStateException("Unexpected registered type: "+type.getClass());
-        }
-
-        protected abstract T visitSpec(RegisteredSpecType type);
-        
-        // TODO beans, others
+    /** Preferred mechanism for defining a bean {@link RegisteredType} */
+    public static RegisteredType bean(String symbolicName, String version, TypeImplementationPlan plan, @Nullable Class<?> superType) {
+        return addSuperType(new BasicRegisteredType(RegisteredTypeKind.BEAN, symbolicName, version, plan), superType);
     }
     
-    public static RegisteredTypeKind getKindOf(RegisteredType type) {
-        return new RegisteredTypeKindVisitor<RegisteredTypeKind>() {
-            @Override protected RegisteredTypeKind visitSpec(RegisteredSpecType type) { return RegisteredTypeKind.SPEC; }
-        }.visit(type);
+    public static RegisteredType spec(String symbolicName, String version, TypeImplementationPlan plan, @Nullable Class<?> superType) {
+        return addSuperType(new BasicRegisteredType(RegisteredTypeKind.SPEC, symbolicName, version, plan), superType);
     }
-    
-    public abstract static class AbstractRegisteredType implements RegisteredType {
 
-        final String symbolicName;
-        final String version;
+    /** returns the {@link Class} object corresponding to the given java type name,
+     * using the cache on the type and the loader defined on the type
+     * @param mgmt */
+    @Beta
+    // TODO should this be on the AbstractTypePlanTransformer ?
+    public static Class<?> loadActualJavaType(String javaTypeName, ManagementContext mgmt, RegisteredType type, RegisteredTypeLoadingContext context) throws Exception {
+        Class<?> result = ((BasicRegisteredType)type).getCache().get(ACTUAL_JAVA_TYPE);
```

Comment:
interesting idea.  have added `@Beta // TODO` on `getCache()` suggesting that, if it ends up that the cache is not wanted elsewhere